README.md
PHP BYTE API Client
========================

A simple wrapper for the BYTE API that requires PHP >= 7.0.

## Requirements

* PHP >= 7.0 with [cURL](http://php.net/manual/en/book.curl.php) extension,

## Installation

The recommended way is using [composer](http://getcomposer.org):

```bash
$ composer require nybbl/byte
```

## Basic usage

```php
use Nybbl\Byte;

$client = new Byte\Client('<my token>');
$response = $client->api('license')->verify('<license code>');
```

## Documentation
* Official [API documentation](https://docs.byte.nybbl.io).

## Contributing

Feel free to make any comments, file issues or make pull requests.

CHANGELOG.md
# Changelog
### Version 0.0.4
---------
* Initial version.

src/Byte/Api/Install.php
<?php
/**
 * Created by PhpStorm.
 * User: Nic
 * Date: 01/11/2018
 * Time: 16:42
 */

namespace Nybbl\Byte\Api;

/**
 * Class Install
 * @package Nybbl\Byte\Api
 */
class Install extends AbstractApi
{
    /**
     * @param string $license
     * @param string $product
     * @param string $ipAddress
     * @param string $domain
     * @param string|null $requestPath
     * @param string|null $serverName
     * @param array $metaData
     * @return mixed
     */
    public function install(
        string $license,
        string $product,
        string $ipAddress,
        string $domain,
        $requestPath = null,
        $serverName = null,
        array $metaData = []
    )
    {
        return $this->post('/install', [
            'license'     => $license,
            'product'     => $product,
            'ipAddress'   => $ipAddress,
            'domain'      => $domain,
            'requestPath' => $requestPath,
            'serverName'  => $serverName,
            'metaData'    => $metaData,
        ]);
    }
}

src/Byte/Api/Version.php
<?php
/**
 * Created by PhpStorm.
 * User: Nic
 * Date: 01/11/2018
 * Time: 16:55
 */

namespace Nybbl\Byte\Api;

/**
 * Class Version
 * @package Nybbl\Byte\Api
 */
class Version extends AbstractApi
{
    /**
     * @param string $id
     * @return mixed
     */
    public function latest(string $id)
    {
        return $this->get(sprintf('products/%s/version/latest', $id));
    }

    /**
     * @param string $id
     * @param string $versionId
     * @return mixed
     */
    public function version(string $id, string $versionId)
    {
        return $this->get(sprintf('products/%s/versions/%s', $id, $versionId));
    }

    /**
     * @param string $id
     * @param string $versionId
     * @param string $compareId
     * @return mixed
     */
    public function compare(string $id, string $versionId, string $compareId)
    {
        return $this->get(sprintf(
            'products/%s/versions/%s/compare/%s',
            $id,
            $versionId,
            $compareId
        ));
    }
}

src/Byte/Api/AbstractApi.php
<?php
/**
 * Created by PhpStorm.
 * User: Nic
 * Date: 01/11/2018
 * Time: 16:26
 */

namespace Nybbl\Byte\Api;

use Curl\Curl;

/**
 * Class AbstractApi
 * @package Nybbl\Byte\Api
 */
abstract class AbstractApi
{
    /** @var string $apiUrl */
    private $apiUrl = 'https://byte.nybbl.io/api/v1';

    /** @var Curl $curlClient */
    private $curlClient;

    /** @var string $accessToken */
    private $accessToken;

    /**
     * AbstractApi constructor.
     * @param string $accessToken
     * @throws \ErrorException
     */
    public function __construct($accessToken = '')
    {
        $this->buildCurlClient();
        $this->accessToken = $accessToken;
    }

    /**
     * @param string $endpoint
     * @param array $data
     * @return mixed
     */
    protected function get(string $endpoint, array $data = [])
    {
        return $this->doApiRequest($endpoint, 'GET', $data);
    }

    /**
     * @param string $endpoint
     * @param array $data
     * @return mixed
     */
    protected function post(string $endpoint, array $data = [])
    {
        return $this->doApiRequest($endpoint, 'POST', $data);
    }

    /**
     * @param string $endpoint
     * @param array $data
     * @return mixed
     */
    protected function put(string $endpoint, array $data = [])
    {
        return $this->doApiRequest($endpoint, 'PUT', $data);
    }

    /**
     * @param string $endpoint
     * @param array $data
     * @return mixed
     */
    protected function patch(string $endpoint, array $data = [])
    {
        return $this->doApiRequest($endpoint, 'PATCH', $data);
    }

    /**
     * @param string $endpoint
     * @param array $data
     * @return mixed
     */
    protected function delete(string $endpoint, array $data = [])
    {
        return $this->doApiRequest($endpoint, 'DELETE', $data);
    }

    /**
     * Sets the Curl client.
     * @throws \ErrorException
     */
    public function buildCurlClient()
    {
        /** @var Curl $curl */
        $curl = new Curl();

        $curl->setUserAgent('nybbl.byte.api.lib');
        $this->setCurlClient($curl);
    }

    /**
     * @param string $endpoint
     * @param string $method
     * @param array $data
     * @return mixed
     */
    private function doApiRequest(string $endpoint, string $method, array $data = [])
    {
        // Build the complete URL to include the API token
        $endpoint = $this->buildURL($endpoint);

        switch (strtolower($method)) {
            case 'get':
                if (!empty($data)) {
                    $this->getCurlClient()->get($endpoint. '&' . http_build_query($data));
                } else {
                    $this->getCurlClient()->get($endpoint);
                }
                break;

            case 'post':
                if (!empty($data)) {
                    $this->getCurlClient()->post($endpoint, $data);
                } else {
                    $this->getCurlClient()->post($endpoint);
                }
                break;

            case 'put':
                if (!empty($data)) {
                    $this->getCurlClient()->put($endpoint, $data);
                } else {
                    $this->getCurlClient()->put($endpoint);
                }
                break;

            case 'patch':
                if (!empty($data)) {
                    $this->getCurlClient()->patch($endpoint, $data);
                } else {
                    $this->getCurlClient()->patch($endpoint);
                }
                break;

            case 'delete':
                if (!empty($data)) {
                    $this->getCurlClient()->delete($endpoint, [], $data);
                } else {
                    $this->getCurlClient()->delete($endpoint);
                }
                break;
        }

        return json_decode(json_encode($this->getCurlClient()->response), true);
    }

    /**
     * Sets the access token.
     * @param string $accessToken
     */
    public function authenticate(string $accessToken)
    {
        $this->accessToken = $accessToken;
    }

    /**
     * @param string $endpoint
     * @return string
     */
    private function buildUrl(string $endpoint): string
    {
        $url = $this->apiUrl . $endpoint;
        $token = '?token=' . $this->accessToken;

        return $url . $token;
    }

    # --------------------------------------------------------------------
    # GETTERS AND SETTERS
    # --------------------------------------------------------------------

    /**
     * @return Curl
     */
    protected function getCurlClient(): Curl
    {
        return $this->curlClient;
    }

    /**
     * @param Curl $curlClient
     * @return $this
     */
    protected function setCurlClient(Curl $curlClient)
    {
        $this->curlClient = $curlClient;
        return $this;
    }

    /**
     * @return string
     */
    public function getApiUrl(): string
    {
        return $this->apiUrl;
    }

    /**
     * @param string $apiUrl
     * @return $this
     */
    protected function setApiUrl(string $apiUrl)
    {
        $this->apiUrl = $apiUrl;
        return $this;
    }

    /**
     * @return string
     */
    public function getAccessToken(): string
    {
        return $this->accessToken;
    }

    /**
     * @param string $accessToken
     * @return $this
     */
    public function setAccessToken(string $accessToken)
    {
        $this->accessToken = $accessToken;
        return $this;
    }
}

src/Byte/Client.php
<?php
/**
 * Created by PhpStorm.
 * User: Nic
 * Date: 01/11/2018
 * Time: 16:26
 */

namespace Nybbl\Byte;

use Nybbl\Byte\Api;
use Nybbl\Byte\Api\AbstractApi;

/**
 * Class Client
 * @package Nybbl\Byte
 */
class Client extends AbstractApi
{
    /**
     * Client constructor.
     * @param string $accessToken
     * @throws \ErrorException
     */
    public function __construct(string $accessToken = '')
    {
        parent::__construct($accessToken);
    }

    /**
     * @param string $name
     * @return Api\Install|Api\License|Api\Product|Api\Version
     * @throws \ErrorException
     * @throws \Exception
     */
    public function api(string $name)
    {
        switch ($name) {
            case 'install':
                return new Api\Install($this->getAccessToken());
                break;

            case 'license':
                return new Api\License($this->getAccessToken());
                break;

            case 'product':
                return new Api\Product($this->getAccessToken());
                break;

            case 'version':
                return new Api\Version($this->getAccessToken());
                break;

            default:
                throw new \Exception('Unknown class! You tried to call: ' . $name);
        }
    }

    /**
     * Proxies $class->product() to $class->api('product')
     *
     * @param string $name
     * @param $args
     * @return Api\Install|Api\License|Api\Product|Api\Version
     * @throws \ErrorException
     */
    public function __call(string $name, $args)
    {
        return $this->api($name);
    }
}